src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp

```diff
@@ -1,139 +1,257 @@
 #include "canvas_messenger.h"
 
+#include <firmware/main/shared_util/constants.h>
+
 #include "util/constants.h"
 
-namespace Util
+using namespace Util;
+
+std::shared_ptr<CanvasMessenger> CanvasMessenger::getInstance()
 {
-    std::shared_ptr<CanvasMessenger> CanvasMessenger::getInstance()
-    {
-        static std::shared_ptr<CanvasMessenger> canvas_messenger(new CanvasMessenger);
-        return canvas_messenger;
-    }
+    static std::shared_ptr<CanvasMessenger> canvas_messenger(new CanvasMessenger);
+    return canvas_messenger;
+}
+
+void CanvasMessenger::initializePublisher(ros::NodeHandle node_handle)
+{
+    this->publisher = node_handle.advertise<thunderbots_msgs::CanvasLayer>(
+        Util::Constants::VISUALIZER_DRAW_LAYER_TOPIC, BUFFER_SIZE);
+}
+
+void CanvasMessenger::publishAndClearAllLayers()
+{
+    // Take ownership of the layers for the duration of this function
+    std::lock_guard<std::mutex> best_known_pass_lock(layers_map_lock);
```

Comment:
I am not sure why `publishAndClearAllLayers` need to be thread safe. Are we planning to call it from multiple threads? Can the main thread call it once every tick? I'm affraid the fact that we need to call it from multiple thread is an indicator of code smell and a more effective implementation of `CanvasMessenger` may be needed. Perhaps looking at limiting layer ownership to one thread to avoid weird visual behavior, and using a commit strategy to only send complete layers to the visualizer. The main thread would be responsible for sending layer data to the visualizer once every tick.

Reply:
We only need to call this function from one thread, but since it touches member variables that might be updated from multiple threads it needs to take a lock to prevent the data from changing while we're publishing it.